keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1186,48 +1188,42 @@ are included with the key provided. E.g.:
 }
 ```
 
-#### Namespace scoped policy binding
-
-For phase 1, policy bindings were only allowed to be cluster scoped. We can
-support namespace scoped policy bindings as follows:
-
-- Add a `NamespacePolicyBinding` resource.
-- If the parameter resource is namespace scoped, it implicitly matches
-  resources only in the namespace it is in, but may further constrain what
-  resources it matches with additional match criteria.
-
-Benefits: Allows policy of a namespace to be controlled from within the
-namespace. As an example, ResourceQuota works this way.
-
-Details to consider:
+#### Per namespace policy params
 
-- Should a policy support both cluster scoped and namespace scoped binding? If
-  so how? It would need two different parameter CRDs (since a CRD must either be
-  cluster scoped or namespace scoped, not both).
+Currently the policies and bindings are only allowed to be cluster scoped.
+We want to support per namespace configuration with namespace scoped param resources.
 
-#### CEL Expression Composition
+(Thanks for the input from @dead2k)
+This sort of mapping allows:
+- A cluster-admin can write a single resource to say, “this is the policy I want in all my namespaces”.
+- If namespace admins can read the param resources, but not write that resource, they can understand the limitations they currently have.
+- A single lenient cluster policy and cluster policybinding can enforce the minimum constraint, and a single cluster policy with a cluster policybinding pointing to a namespace level param can further restrict.
 
-##### Use Cases
-
-###### Code re-use for complicated expressions
+A new optional field `namespaceParamRef` could be added inside ValidatingAdmissionPolicyBinding to support such use case.
+In contrast, a namespace scoped policybinding will require creation and maintenance of both policybindings and parameters
+in every namespace to enforce the policy itself, versus the single policybinding and many parameters.
 
-A CEL expression may not be computationally expensive, but could still be
-intricate enough that copy-pasting could prove to be a bad decision later on
-in time. With the addition of the `messageExpression` field, more copy-pasting
-is expected as well. If a sufficiently complex expression ended up copy-pasted everywhere,
-and then needs to be updated somehow, it will need that update in every place
-it was copy-pasted. A variable, on the other hand, will only need to be updated
-in one place.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "demo-binding-test.example.com"
+spec:
+  policyName: "demo-policy.example.com"
+  namespaceParamRef:
+    name: "param-resource.example.com"
```

Comment:
I think we want an option of name or label selector here to mean, "this particular one" or  "honor all the parameter resources that match this label selector in that namespace" to allow slightly looser coupling.   That allows evolution of namespace names using a kustomize style  hash in each namespace with a single ValidatingAdmissionPolicyBinding working against many namespaces.

Reply:
Added. I haven't showed a kustomize example, but I have shown one with a label selector along with a summary of the behavior if the selector matches multiple resources.

Reply:
> Added. I haven't showed a kustomize example, but I have shown one with a label selector along with a summary of the behavior if the selector matches multiple resources.

Commit https://github.com/kubernetes/enhancements/pull/3949/commits/057a9c31ce34e277f5dadb27d52eb1261053f027 appears to be missing from the diff.  Conflicting push somewhere?

It's exactly what I'm thinking though.